plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ConstantPopulatingPageSource.java

```diff
@@ -0,0 +1,180 @@
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ConstantPopulatingPageSource
+        implements ConnectorPageSource
+{
+    private final ConnectorPageSource delegate;
+    private final int size;
+    private final Map<Integer, Block> constantColumns;
+    private final Map<Integer, Integer> sourceChannelToTargetChannel;
+
+    public ConstantPopulatingPageSource(ConnectorPageSource delegate, List<ColumnType> columns)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        requireNonNull(columns, "columns is null");
+        this.size = columns.size();
+
+        ImmutableMap.Builder<Integer, Block> constantColumns = ImmutableMap.builder();
+        ImmutableMap.Builder<Integer, Integer> sourceChannelToTargetChannel = ImmutableMap.builder();
+
+        for (int targetChannel = 0; targetChannel < columns.size(); targetChannel++) {
+            ColumnType column = columns.get(targetChannel);
+            if (column instanceof ConstantColumn) {
+                constantColumns.put(targetChannel, ((ConstantColumn) column).getValue());
+            }
+            else if (column instanceof DelegateColumn) {
+                sourceChannelToTargetChannel.put(((DelegateColumn) column).getSourceChannel(), targetChannel);
+            }
+            else {
+                throw new IllegalArgumentException("Unknown ColumnType: " + column);
+            }
+        }
+
+        this.constantColumns = constantColumns.buildOrThrow();
+        this.sourceChannelToTargetChannel = sourceChannelToTargetChannel.buildOrThrow();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public OptionalLong getCompletedPositions()
+    {
+        return delegate.getCompletedPositions();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page delegatePage = delegate.getNextPage();
+
+        if (delegatePage == null) {
+            return null;
+        }
+
+        Block[] blocks = new Block[size];
+        for (int sourceChannel = 0; sourceChannel < delegatePage.getChannelCount(); sourceChannel++) {
+            blocks[sourceChannelToTargetChannel.get(sourceChannel)] = delegatePage.getBlock(sourceChannel);
+        }
+
+        for (Map.Entry<Integer, Block> constantColumnEntry : constantColumns.entrySet()) {
+            blocks[constantColumnEntry.getKey()] = new RunLengthEncodedBlock(constantColumnEntry.getValue(), delegatePage.getPositionCount());
```

Comment:
There is an inherent assumption here that `sourceChannelToTargetChannel.get(sourceChannel)` will always be different from `constantColumnEntry.getKey()` . Should we validate that in the constructor?